### DISCUSSION TITLE
Template Strings

### DISCUSSION THREAD

**Discussion by LucaGabi** [dotnet/csharplang#2216] [Language Ideas]
Hi,

In Xtend a java based language (transpiles to java code) there is a feature called string templates that allows easy code generation with logic like javascript mustache template engine but only in the java context and looks like this:

![image](https://user-images.githubusercontent.com/37976898/52733326-2230c980-2fcb-11e9-8bb3-1793ee00c55b.png)

![image](https://user-images.githubusercontent.com/37976898/52734289-763cad80-2fcd-11e9-9c21-1d7d9856ad2d.png)

It is often a case when you need to generate class models or other type of repetitive code (even GraphQL entites and not only) that is why I thought this to be a good feature because it can use all the context of dotnet framework and custom referenced dlls. 

https://www.eclipse.org/xtend/documentation/203_xtend_expressions.html#templates

[T4 text template](https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates?view=vs-2017) - was a similar tool but now it is not supported in dotnet core

INITIAL OLD:
   I find very useful the string interpolation feature and I was thinking even more useful would be a string template feature like Kotlin has:

It would be of great help in code generation.

**Comment by HaloFour** [comment]
C# already has verbatim interpolated strings.  Your example seems to just be a raw string, though, and has nothing to do with interpolation.

**Comment by Logerfo** [comment]
You can achieve something similar of what I think you want with the following code:
```c#
public static string Template(Func<IEnumerable<string>> func)
{
    var builder = new StringBuilder();
    foreach (var item in (func ?? throw new ArgumentNullException(nameof(func))).Invoke())
        builder.Append(item);

    return builder.ToString();
}

void main()
{
    IEnumerable<string> _str()
    {
        var i = 0;
        while (i <= 5)
        {
            yield return i.ToString();
            yield return Environment.NewLine;
        }
    }
    var str = Template(_str);
    Console.WriteLine(str);
}
```

**Comment by LucaGabi** [comment]
Maybe I was not clear enough: I know about the string interpolation feature in C# the new feature is about string templates: """  """ this is the operator in Kotlin and it allows for multi line strings and expressions (if,for,do, lambdas, method calls etc.) right in the string. 

I think this would be useful to be able to generate code.

**Comment by HaloFour** [comment]
Your mention about string interpolation is what makes this unclear.  Kotlin only allows expressions in string templates, just like C#.  You can't have assignments or loops.  Your example produces a String that happens to contain that Kotlin code, it doesn't execute anything.

If your proposal is about producing a string that contains such code and has nothing to do with interpolation, then see the following proposal.  It's based more on C++'s implementation of raw strings rather than Python's or Kotlin's, but the premise is the same.

https://github.com/dotnet/csharplang/issues/89

**Comment by LucaGabi** [comment]
Sorry, look I'll post a reference picture:
![image](https://user-images.githubusercontent.com/37976898/52661399-9eafa380-2f0a-11e9-937c-8ea40414a77b.png)
![image](https://user-images.githubusercontent.com/37976898/52661521-ecc4a700-2f0a-11e9-8ceb-9919c53bbda9.png)

**Comment by HaloFour** [comment]
I'm not sure what what's supposed to be?  That doesn't look like Kotlin syntax.  What language lets you use guillemets for interpolation?

**Comment by LucaGabi** [comment]
Yes, you're right: the blue text is string output as it is, the « character is the equivalent of  { in c# interpolated string the » charcacter is the } in the c# interpolated string. The purple colored text is part of an instruction or expression as reserved word. The last picture shows a function definition with an implicit return statement of the evaluated string template. The srting template is started with (3) ''' and also end with (3) ''' anything that is enclosed in « » will be evaluated, it's like a mustache template in javascript.

Can such a feature go in C# ?

The language is an extention of java called Xtend.
https://www.eclipse.org/xtend/

**Comment by HaloFour** [comment]
You can use interpolated verbatim strings:

```cs
var message = $@"
    Hello {name}!
";
```

You have to escape double-quotes by doubling them up, but otherwise it's a raw string.  Combining raw strings with templates seems like it would defeat the purpose as you end up having to escape something.

**Comment by LucaGabi** [comment]
I can't see how this would help with not having to split and concatenate the string over and over to keep things readable when there is a need to have an output value for a logical evaluation or loop iteration for  multiple values. It it's not about not being able to have the result but about how easy and readable it is.

I still believe this would be good feature to have. There are many situations when code generation can be used to shorten the time and possible errors of repetitive code.

The posted example dose not support statements evaluation of any kind: if, while, for.
I'd be really happy for this to be considered.

Thanks anyway for all the answers.

**Comment by HaloFour** [comment]
@LucaGabi 

I can't imagine that such a use case would be very common and it sounds like it would be quite complicated to implement, especially supporting arbitrary block of code within the string.  It's basically rebuilding the MVC razor engine within the language.

**Comment by scalablecory** [comment]
I don't know if this is widely applicable enough for integration into C#.

Or even appropriate given the current landscape: generally the industry has moved away from putting templating directly into code, and instead pushed it out to templating frameworks like Razor. This gives a better separation of concerns.

(Regardless, if you want to pursue this I encourage you to edit your original comment to be clearer -- I had to read down to your latest comment to really understand what you were after.)

**Comment by markusamshove** [comment]
The special thing about the strings in Xtend (in the screenshots) is actually the greyspace. 

Such a string generates a string builder in java which is aware of indentation.
It means that only parts of the string get generated which are highlighted by the greyspace. 
Every character that is on a new line and has no grey space to the left will be at the start of the line.

In that way It is different than `@""`.

For Xtend it makes sense to have such a construct, because the initial purpose of the language was to generate code from a domain specific language.


To have the same with Kotlin or groovy you would call stripMargin on the string.

It would be a cool addition to C#, but only really beneficial if there is IDE support for it like I'm Xtend.

[Here](https://www.eclipse.org/xtend/documentation/203_xtend_expressions.html#templates) is the info from their docs (scroll down to whitespace handling)

**Comment by TonyValenti** [comment]
This makes me think of Razor.

**Comment by LucaGabi** [comment]
> This makes me think of Razor.

More like T4 Text Template rather ..
